aten/src/ATen/native/cuda/group_norm_kernel.cu

```diff
@@ -40,11 +73,70 @@ __global__ void RowwiseMomentsCUDAKernel(
   using WelfordOp =
       WelfordOps<T_ACC, T_ACC, int64_t, thrust::pair<T_ACC, T_ACC>>;
 
+  const int G = C / num_ch_group;
+
   const int64_t i = blockIdx.x;
+  const int64_t group_channels = C / G;
+  const int64_t group_idx = i % G;
+  const int64_t group_offset = group_idx * group_channels;
+  const int64_t batch_idx = i / G;
+
+  const int64_t batch_offset = batch_idx * H * W * C;
+
   WelfordOp welford_op = {/*correction=*/0, /*take_sqrt=*/false};
   WelfordType val(0, 0, 0, 0);
-  for (int64_t j = threadIdx.x; j < N; j += blockDim.x) {
-    const int64_t index = i * N + j;
+  for (int64_t j = threadIdx.x; j < group_span; j += blockDim.x) {
+    const int64_t hw_idx = j;
+    const int64_t hw_offset = hw_idx * C;
+    const int64_t index = batch_offset + hw_offset + group_offset;
+
+    for (int64_t c = 0; c < group_channels; c++) {
+      val = welford_op.reduce(val, static_cast<T_ACC>(X[index + c]), index + c);
+    }
```

Comment:
This kernel uses a different indexing strategy that will work for rNHWC tensors and uses welfords algorithm.